src/core_test/mpas_test_core_stream_list.F

```diff
@@ -0,0 +1,638 @@
+      if (ierr /= MPAS_STREAM_LIST_NOERR) then
+         err = err + 1
+      end if
+      if (.not. associated(removed2)) then
+         err = err + 1
+      end if
+      if (trim(removed2%name) /= 'stream2') then
+         err = err + 1
+      end if
+
+      ! Remove from the end
+      call MPAS_stream_list_remove(list, 'stream3', removed3, ierr=ierr)
+      if (ierr /= MPAS_STREAM_LIST_NOERR) then
+         err = err + 1
+      end if
+      if (.not. associated(removed3)) then
+         err = err + 1
+      end if
+      if (trim(removed3%name) /= 'stream3') then
+         err = err + 1
+      end if
+
+      ! Remove from the beginning
+      call MPAS_stream_list_remove(list, 'stream1', removed1, ierr=ierr)
+      if (ierr /= MPAS_STREAM_LIST_NOERR) then
+         err = err + 1
+      end if
+      if (.not. associated(removed1)) then
+         err = err + 1
+      end if
+      if (trim(removed1%name) /= 'stream1') then
+         err = err + 1
+      end if
+
+      call MPAS_stream_list_destroy(list)
+      deallocate(removed1)
+      deallocate(removed2)
+      deallocate(removed3)
+   end subroutine mpas_test_remove_existing_streams
+   !**************************************************************************************
+   !  Subroutine mpas_test_insert_non_adjacent_duplicate
+   !
+   !> \brief   Test inserting a non-adjacent duplicate of the first item added to the list.
+   !>
+   !> \details This subroutine verifies that inserting a duplicate of the first stream
+   !>          added to an MPAS stream list results in the correct duplicate error
+   !>          code when the duplicate is not inserted immediately after the original.
+   !>          It does so by inserting two unique streams into the list, then attempting
+   !>          to insert the first stream again. This confirms that duplicate detection
+   !>          works for non-adjacent duplicates in insertion order.
+   !>
+   !> \param err  The error code that indicates the result of the test.
+   !**************************************************************************************
+   subroutine mpas_test_insert_non_adjacent_duplicate(err)
+      integer, intent(out) :: err
+      type(MPAS_stream_list_type), pointer :: list
+      type(MPAS_stream_list_type), pointer :: s1, s2
+      integer :: ierr
+
+      err = 0
+      call MPAS_stream_list_create(list)
+
+      allocate(s1)
+      s1%name = 'stream1'
+      allocate(s2)
+      s2%name = 'stream2'
+
+      call MPAS_stream_list_insert(list, s1, ierr)
+      if (ierr /= MPAS_STREAM_LIST_NOERR) then
+         err = err + 1
+      end if
+
+      call MPAS_stream_list_insert(list, s2, ierr)
+      if (ierr /= MPAS_STREAM_LIST_NOERR) then
+         err = err + 1
+      end if
+
+      call MPAS_stream_list_insert(list, s1, ierr)
+      if (ierr /= MPAS_STREAM_LIST_DUPLICATE) then
+         err = err + 1
+      end if
+      if (.not. associated(s1%next)) then
```

Comment:
Without knowing the details of how the `MPAS_stream_list_insert` routine is implemented (e.g., do we add new items at the head of the list, or at the tail of the list?), it might not be obvious to the reader why we are testing that `s1%next` is associated here. I think an comment would be helpful here.